bin/cloudworker.js

```diff
@@ -44,20 +46,42 @@ function run (file, wasmBindings) {
   Object.assign(bindings, wasmBindings)
 
   const opts = {debug: program.debug, enableCache: program.enableCache, bindings: bindings}
-  const server = new Cloudworker(script, opts).listen(program.port)
+  let server = new Cloudworker(script, opts).listen(program.port)
 
   console.log(`Listening on ${program.port}`)
 
   let stopping = false
+  let reloading = false
+
+  if (program.watch) {
+    fs.watchFile(fullpath, () => {
+      reloading = true
+      console.log('Changes to the worker script detected - reloading...')
+
+      server.close(() => {
+        if (stopping) return
+
+        reloading = false
+        console.log('Successfully reloaded!')
+
+        server = new Cloudworker(utils.read(fullpath), opts).listen(program.port)
+      })
+    })
+  }
+
   function shutdown () {
     if (stopping) return
 
     stopping = true
     console.log('\nShutting down...')
-    server.close(() => {
-      console.log('Goodbye!')
-      process.exit(0)
-    })
+    server.close(terminate)
+
+    if (reloading) server.on('close', terminate)
```

Comment:
I _think_ this should be moved right below `stopping = true` and a return statement should be added to the if block. `reloading` will only be true if `close` has already been invoked in the `watchFile` callback (e.g. the server is still shutting down so has yet to call the callback). If we set `stopping` to true, the callback for the `close` that originated in `watchFile` will return early and `terminate` will run and exit the process.

Reply:
My reasoning for the way I had done it was for consistent logging regardless as to whether it was running as a watcher or not it would output the same logs.

See my suggestions now, if you agree with them then I'll submit an update! :)

Reply:
I think I might have been mistaken about the race condition. Let me play with  this PR. It may be fine as-is.

Reply:
Yeah, I was trying to account for the odd case where somebody may terminate just as the server has successfully closed while reloading.

The race condition could occur if line 62 had been passed, hence why I added the event listener.